test/__snapshots__/loader.test.js.snap

```diff
@@ -733,6 +733,20 @@ Array [
     "./index-loader-syntax.css",
     ".a {
   color: red;
+}",
+    "",
+  ],
+  Array [
+    "button.modules.css!=!./index-loader-syntax-sass.css",
+    ".baz {
+  width: 5px;
+}",
```

Comment:
note they aren't hashed

Reply:
Why do not use `index-loader-syntax-sass.modules.css`? Because request here is `index-loader-syntax-sass.css`

Reply:
that wouldn't illustrate the problem. css-loader only looks at the resourcePath, not the matchResource. The whole goal of using this syntax is as an alternative to virtual files. **ideally** i'd want this to work in the case where there is no `resourcePath`

Reply:
e.g. having this `button.module.scss!=!./base64-loader/index.js?[ident]` work, at least from what Tobias suggested in the other issue. adding a resource to the end is just a hack to get something working

Reply:
I think we can't do it, because `button.module.scss` can be any format...

Reply:
that's true of any file you point a loader at...it's confusing to me that 'button.module.scss' is used to determine what loaders apply to this resource, but it's not passed to the loaders themselves.

like the goal of the `!=!` syntax seems be for *telling* you what format the file is. 

again the goal would be for `button.module.scss!=!./base64-loader/index.js?[ident]` to work *as is* with css-loader. This syntax _does_ work in webpack, it treats the above as a `.module.scss` file, the problem is css-loader doesn't (because it has no info)

Reply:
Please describe what you want to solve

Reply:
i've describe what i want to solve many times now in this issue: https://github.com/webpack/webpack/issues/11074

please let me know if there is something else i can be clearer on

Reply:
in a nutshell I want to turn: 

```js
const styles = css`
  .foo {
    color: red
  }
`
```
into 

```js
styles = require('./generated-file-that-doesn't-exist-on-disk.module.css')
```

I was using virtual modules to do this, @sokra said, that was a bad idea and i should turn it into:

```js
styles = require('.styles.module.css!=!content-from-query-loader?"foo { color: red;}"!')
```

I am now trying to make that work with css-loader

Reply:
Why do not use `styles = require('styles.module.css!=!content-from-query-loader?"foo { color: red;}"!./original-file.modules.js')`?

So we got `original` file when move content to you loader where you do transformation, for example - extract CSS, when we move it to `styles.module.css` and apply CSS loader?

Reply:
the JS file is never named `.modules.js`, since that's not a common naming pattern. I can't really tell users or my team to name their React component files `.module.js`, and it's  not possible if the original file is a .svelte, or .vue file or other mixed style format

Reply:
hm, sounds reasonable, give me time to look at this

Reply:
It might be the issue is that webpack needs to expose matchResource to loaders, so they can decide which to use

Reply:
uncovered another subtle problem with using the original file as the resource with css modules:

multiple virtual files from one original gives css-loader different content with the same resource. may affect caching but also affects how class names are hashed,

```js
// becomes button.module.css!=!./my-loader/!.button.js
const button = css`
    .foo {
     color: blue
   }
`
// becomes button2.module.css!=!./my-loader/!.button.js
const button2 = css`
   .foo {
     color: red
   }
`
```

The above, would produce the same hash for each `.foo` (since the same file is used to generate the local name) breaking the encapsulation of the style

Reply:
due this https://github.com/webpack-contrib/css-loader/blob/master/src/utils.js#L73

Reply:
What about `button2.module.css?modules=true!=!./my-loader/!.button.js`?

Reply:
You can implement `modules` on own side

Reply:
Like reimplement it? I could do that, but it's a lot of work. The hope/goal was to use the users existing config. If we need to reimplement css-loader, I can avoid all of this and write the whole thing as a plugin probably but that's more config and more duplicated work. It also doesn't solve the problem for anyone else trying to use virtual modules in this way, like svelte

Reply:
the real hard part though of that plain is when there are things in front of css-loader:

```js
button2.module.css!=!./my-loader?modules=true!.button.js
```
resolves fully to:
```
css-loader!sass-loader!my-loader?modules!./button.js
```

`my-loader` can't do modules at it's spot because Sass still needs to run. my loader also can't enable modules on the lower css-loader because it would change the option for all files. The only thing it can is something like: https://github.com/webpack/webpack/issues/10350 which i say works great, but have since realized *does not work* well at scale for a few, hard to describe, reasons that causes styles to be included multiple times and out of order, due to module `id`'s being (correctly) different between the plain require and compiled requires